packages/lib/src/acting-io.ts

```diff
@@ -0,0 +1,170 @@
+ * @param id The id of the item to act on, or undefined if the user should be asked to select one.
+ * @param listFunction A function that returns a list of all items which will be used to build the
+ *   list of the user to choose from when `id` is undefined.
+ * @param actionFunction The function that performs the action.
+ * @param getIdFromUser A function that queries the user into a for an item given a list of them.
+ *
+ * @returns a tuple containing the id of the item acted on and the result of the action function.
+ */
+export async function selectAndActOnGeneric<ID, O, L>(command: SelectingCommand<L>,
+		id: ID | undefined, listFunction: ListDataFunction<L>, actionFunction: ActionFunction<ID, O>,
+		getIdFromUser: IdRetrievalFunction<ID, L>): Promise<[ID, O]> {
+	let computedId: ID
+	if (id) {
+		computedId = id
+	} else {
+		const list = await outputList(command, listFunction, true)
+		if (list.length === 0) {
+			// Nothing was found; user was already notified.
+			command.exit(0)
+		}
+		computedId = await getIdFromUser(command.primaryKeyName, list)
+	}
+	const updatedItem = await actionFunction(computedId)
+	return [computedId, updatedItem]
+}
+selectActOnAndOutputGeneric.flags = outputList.flags
+
+/**
+ * Process a command that selects and item (e.g. a device), performs some action on that device
+ * and then simply states that the action has completed. This is useful for actions that don't
+ * return data, like deleting an item.
+ *
+ * @param command The command itself which conforms to the `SelectingCommand` type.
+ * @param id The id of the item to act on, or undefined if the user should be asked to select one.
+ * @param listFunction A function that returns a list of all items which will be used to build the
+ *   list of the user to choose from when `id` is undefined.
+ * @param actionFunction The function that performs the action.
+ * @param successMessage The message that should be displayed upon successful completion of the
+ *   action. (Successful completion is defined as `actionFunction` not throwing any exceptions.)
+ *
+ * @returns the id selected and acted on by `actionFunction`
+ */
+export async function selectAndActOn<L>(command: SelectingCommand<L>,
+		id: string | undefined, listFunction: ListDataFunction<L>,
+		actionFunction: ActionFunction<string, void>, successMessage: string): Promise<string> {
+	const [computedId] = await selectAndActOnGeneric<string, void, L>(command, id, listFunction,
+		actionFunction, newStringGetIdFromUser)
+	process.stdout.write(`${successMessage.replace('{{id}}', JSON.stringify(computedId))}\n`)
+	return computedId
+}
+
+
+export type SelectingOutputCommand<O, L> = SelectingCommand<L> & CommonOutputProducer<O>
+
+/**
+ * Process a command that selects an item (e.g. a device), performs some action on that device
+ * and then outputs the results. Unless your ids are not simple strings, you probably want to
+ * use `selectActOnAndOutput` below instead.
```

Comment:
Do we have any use cases where ids aren't simple strings? Maybe we could just not export this.

Reply:
Capabilities have a two part id with a string and a version. It's possible that might change and then we can remove this but for now we do.